Tests/CatalogItems/New-RsDataSource.Tests.ps1

```diff
@@ -0,0 +1,184 @@
+        It "Should be a new data source" {
+            {Get-RsDataSource -Path $dataSourcePath } | Should not throw
+        }
+        # Removing folders used for testing
+        Remove-RsCatalogItem -RsFolder $dataSourcePath
+    }
+
+     Context "Unsupported RsDataSource Extension validation"{
+        # Declare datasource Name, Extension, CredentialRetrieval, and DataSource path.
+        $dataSourceName = 'SutDataSurceExtensionException' + [guid]::NewGuid()
+        $extension = 'InvalidExtension'
+        $credentialRetrieval = 'Integrated'
+        $dataSourcePath = '/' + $dataSourceName
+        It "Should throw an exception when datasource is failed to be create" {
+             { New-RsDataSource -RsFolder '/' -Name $dataSourceName -Extension $extension -CredentialRetrieval $credentialRetrieval } | Should throw 
+             { Get-RsDataSource -Path $dataSourcePath } | Should throw
+        }
+    }
+
+    Context "STORE credential validation" {
+        # Declare datasource Name, Extension, CredentialRetrieval, and DataSource path.
+        $dataSourceName = 'SutDataSurceStoreException' + [guid]::NewGuid()
+        $extension = Get-ExistingExtension
+        $credentialRetrieval = 'STORE'
+        $dataSourcePath = '/' + $dataSourceName
+        It "Should throw an exception when Store credential retrieval are given without providing credential" {
+            { New-RsDataSource -RsFolder '/' -Name $dataSourceName -Extension $extension -CredentialRetrieval $credentialRetrieval } | Should throw 
+            { Get-RsDataSource -Path $dataSourcePath } | Should throw
+        }
+    }
+
+    Context "Create RsDataSource with Data Source Credentials" {
+        # Declare datasource Name, Extension, CredentialRetrieval, and DataSource path.
+        $dataSourceName = 'SutDataSurceCredentials' + [guid]::NewGuid()
+        $extension = Get-ExistingExtension
+        $credentialRetrieval = 'None'
+        $dataSourcePath = '/' + $dataSourceName
+        # Creation of PSCredentials
+        $password ='MyPassword'
+        $user = [System.Security.Principal.WindowsIdentity]::GetCurrent().Name
+        $dataSourceCredentials = Create-PSCredential -User $user -Password $password
+        New-RsDataSource -RsFolder '/' -Name $dataSourceName -Extension $extension -CredentialRetrieval $credentialRetrieval -DataSourceCredential $dataSourceCredentials
+        It "Should be a new data source" {
+            {Get-RsDataSource -Path $dataSourcePath } | Should not throw
+        }
+        # Removing folders used for testing
+        Remove-RsCatalogItem -RsFolder $dataSourcePath
+    }
+
+    Context "Create RsDataSource with ImpersonateUser Parameter"{
+        # Declare datasource Name, Extension, CredentialRetrieval, and DataSource path.
+        $dataSourceName = 'SutDataSurceImpersonateUser' + [guid]::NewGuid()
+        $extension = Get-ExistingExtension
+        $credentialRetrieval = 'None'
```

Comment:
Should this be None or Store? ImpersonateUser doesn't really get used unless you specify Store.